<a href="https://colab.research.google.com/github/angelvaz20/Colab_files/blob/main/Lab_11_XGBM_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Names: Angel Vasquez and Dillan Andrews

Dataset: https://www.kaggle.com/datasets/neuromusic/avocado-prices


In [ ]:
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [ ]:
df = pd.read_csv('avocado.csv', index_col=0)
df

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015,Albany
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015,Albany
2,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015,Albany
3,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015,Albany
4,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015,Albany
...,...,...,...,...,...,...,...,...,...,...,...,...,...
7,2018-02-04,1.63,17074.83,2046.96,1529.20,0.00,13498.67,13066.82,431.85,0.0,organic,2018,WestTexNewMexico
8,2018-01-28,1.71,13888.04,1191.70,3431.50,0.00,9264.84,8940.04,324.80,0.0,organic,2018,WestTexNewMexico
9,2018-01-21,1.87,13766.76,1191.92,2452.79,727.94,9394.11,9351.80,42.31,0.0,organic,2018,WestTexNewMexico
10,2018-01-14,1.93,16205.22,1527.63,2981.04,727.01,10969.54,10919.54,50.00,0.0,organic,2018,WestTexNewMexico


In [ ]:
print(f"Total samples: {len(df)}")
print(f"Number of columns: {len(df.columns)}")
print(f"\nFeature names:")
for i, col in enumerate(df.columns, 1):
  print(f" {i}. {col}")

Total samples: 18249
Number of columns: 13

Feature names:
 1. Date
 2. AveragePrice
 3. Total Volume
 4. 4046
 5. 4225
 6. 4770
 7. Total Bags
 8. Small Bags
 9. Large Bags
 10. XLarge Bags
 11. type
 12. year
 13. region


In [ ]:
print(f"\nTarget variable statistics (Average price in avocado):")
print(f" Minimum: ${df['AveragePrice'].min():2f}")
print(f" Maximum: ${df['AveragePrice'].max():2f}")
print(f" Mean: ${df['AveragePrice'].mean():2f}")
print(f" Median: ${df['AveragePrice'].median():2f}")


Target variable statistics (Average price in avocado):
 Minimum: $0.440000
 Maximum: $3.250000
 Mean: $1.405978
 Median: $1.370000


In [ ]:
# Prepare and clean data
df_clean = df.drop(columns=['Date', 'region'])

df_clean['type'] = df_clean['type'].map({'conventional':0, 'organic':1})

print(f"Cleaned dataset shape: {df_clean.shape}")
print(f"\Missing values per column:")
print(df_clean.isnull().sum())


Cleaned dataset shape: (18249, 11)
\Missing values per column:
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
dtype: int64


<>:7: SyntaxWarning: invalid escape sequence '\M'
<>:7: SyntaxWarning: invalid escape sequence '\M'
/tmp/ipykernel_23551/1288357049.py:7: SyntaxWarning: invalid escape sequence '\M'
  print(f"\Missing values per column:")


In [ ]:
# Define features (X) and target (y)
# Target: AveragePrice — the price we want to predict
# Features: everything else (volume, bag sizes, type, year)
X = df_clean.drop(columns=['AveragePrice']).values
y = df_clean['AveragePrice'].values
feature_names = df_clean.drop(columns=['AveragePrice']).columns.tolist()

print(f"\nFeature names used for prediction:")
for i, name in enumerate(feature_names, 1):
    print(f"  {i}. {name}")

print(f"\nX shape: {X.shape}  |  y shape: {y.shape}")


Feature names used for prediction:
  1. Total Volume
  2. 4046
  3. 4225
  4. 4770
  5. Total Bags
  6. Small Bags
  7. Large Bags
  8. XLarge Bags
  9. type
  10. year

X shape: (18249, 10)  |  y shape: (18249,)


In [ ]:
# Splitt data into training and testing sets
# Split 80% training testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(f"Total dataset size: {len(X)} samples")
print(f"Training set size: {len(X_train)} sample (80%)")
print(f"testing set size: {len(X_test)} samples (20%")

Total dataset size: 18249 samples
Training set size: 14599 sample (80%)
testing set size: 3650 samples (20%


In [ ]:
#Create and train the XGBOOST regressor

model = xgb.XGBRegressor(
    max_depth=4,
    n_estimators=100,
    learning_rate=0.1,
    objective='reg:squarederror',
    random_state=42,
    subsamples=0.8,
    colsample_bytree=0.8
)

print("Model parameters:")
print(f"  - Max tree depth: {model.max_depth}")
print(f"  - Number of trees: {model.n_estimators}")
print(f"  - Learning rate: {model.learning_rate}")
print(f"  - Subsample ratio: {model.learning_rate}")
print(f" Colssample bytree: {model.colsample_bytree}")

Model parameters:
  - Max tree depth: 4
  - Number of trees: 100
  - Learning rate: 0.1
  - Subsample ratio: 0.1
 Colssample bytree: 0.8


In [ ]:
# Train the model
model.fit(X_train, y_train, verbose=False)  # verbose=False to reduce output

/usr/local/lib/python3.12/dist-packages/xgboost/training.py:200: UserWarning: [21:30:00] WARNING: /__w/xgboost/xgboost/src/learner.cc:782: 
Parameters: { "subsamples" } are not used.

  bst.update(dtrain, iteration=i, fobj=obj)


XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=0.8, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.1, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=4,
             max_leaves=None, min_child_weight=None, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=100,
             n_jobs=None, num_parallel_tree=None, ...)

In [ ]:
# Make prediction
y_train_pred = model.predict(X_train)
y_test_pred = model.predict(X_test)

In [ ]:
# Evaluate the model

# --- Compute Metrics ---
train_rmse = np.sqrt(mean_squared_error(y_train, y_train_pred))
test_rmse  = np.sqrt(mean_squared_error(y_test,  y_test_pred))
train_mae  = mean_absolute_error(y_train, y_train_pred)
test_mae   = mean_absolute_error(y_test,  y_test_pred)
train_r2   = r2_score(y_train, y_train_pred)
test_r2    = r2_score(y_test,  y_test_pred)

print("\n" + "=" * 70)
print("MODEL EVALUATION")
print("=" * 70)

print("\nTraining Set Performance:")
print(f"  RMSE (Root Mean Squared Error): ${train_rmse:.4f}")
print(f"  MAE  (Mean Absolute Error):     ${train_mae:.4f}")
print(f"  R²   (R-Squared Score):          {train_r2:.4f}")

print("\nTest Set Performance:")
print(f"  RMSE (Root Mean Squared Error): ${test_rmse:.4f}")
print(f"  MAE  (Mean Absolute Error):     ${test_mae:.4f}")
print(f"  R²   (R-Squared Score):          {test_r2:.4f}")

print(f"\nInterpretation:")
print(f"  The model explains {test_r2 * 100:.1f}% of the variance in avocado prices.")

# Show sample predictions vs actual prices
print("\nSample predictions vs actual prices (first 10 test samples):")
print(f"{'Actual Price':>15} {'Predicted Price':>17} {'Difference':>12}")
print("-" * 46)
for actual, predicted in zip(y_test[:10], y_test_pred[:10]):
    diff = predicted - actual
    print(f"  ${actual:>10.2f}     ${predicted:>12.2f}     ${diff:>+9.2f}")


MODEL EVALUATION

Training Set Performance:
  RMSE (Root Mean Squared Error): $0.2118
  MAE  (Mean Absolute Error):     $0.1579
  R²   (R-Squared Score):          0.7240

Test Set Performance:
  RMSE (Root Mean Squared Error): $0.2262
  MAE  (Mean Absolute Error):     $0.1695
  R²   (R-Squared Score):          0.6816

Interpretation:
  The model explains 68.2% of the variance in avocado prices.

Sample predictions vs actual prices (first 10 test samples):
   Actual Price   Predicted Price   Difference
----------------------------------------------
  $      0.82     $        1.10     $    +0.28
  $      0.97     $        1.08     $    +0.11
  $      1.44     $        1.56     $    +0.12
  $      0.97     $        1.06     $    +0.09
  $      1.45     $        1.48     $    +0.03
  $      1.29     $        1.58     $    +0.29
  $      2.03     $        1.74     $    -0.29
  $      0.94     $        1.04     $    +0.10
  $      2.45     $        1.91     $    -0.54
  $      1.83     $   

# Summary:
This lab uses the Avocado Prices, which contains over 18,000 weekly retail records of Hass avocado sales collected across major U.S. cities from 2015 to 2018. Each row represents one week of sales data for a specific region and includes information such as the total volume of avocados sold, the breakdown of sales by PLU code (4046, 4225, and 4770 refer to different sizes of Hass avocados), bag sizes (small, large, and extra-large), whether the avocados were conventional or organic, and the year. The goal of the project is to predict the average retail price of avocados using these features, making it a regression problem, since price is a continuous numerical value rather than a category.

To build the model, we used XGBoost (Extreme Gradient Boosting). The code follows a clean five-step pipeline: load and explore the data, prepare and clean it (which included dropping the date and region text columns and converting the avocado type into a number), split it 80/20 into training and testing sets, train the XGBRegressor with hyperparameters such as max depth, learning rate, subsample, and colsample_bytree, and finally evaluate the model using RMSE, MAE, and R².

The most interesting finding from the model is the feature importance result: the single most influential factor in predicting avocado price is the type, whether the avocado is conventional or organic, which accounted for roughly 85% of the model's decision making. This makes strong real-world sense, since organic avocados consistently sell at a premium compared to conventional ones. The model achieved an R² of about 0.68 on the test set, meaning it explained approximately 68% of the variation in avocado prices, with an average prediction error of around 23 cents. While there is room for improvement through further tuning or adding the region back as an encoded feature, the model successfully captures the dominant pricing patterns in the data and demonstrates how XGBoost can be applied to a real-world economic prediction problem.

